server/mailet/integration-testing/src/test/java/org/apache/james/mailets/ForwardLoopIntegrationTest.java

```diff
@@ -0,0 +1,197 @@
+    private static final Username CEDRIC = Username.of("cedric@" + DEFAULT_DOMAIN);
+    private static final MailRepositoryUrl CUSTOM_REPOSITORY = MailRepositoryUrl.from("memory://var/mail/custom/");
+    public static final Rule.ConditionGroup CONDITION_GROUP = Rule.ConditionGroup.of(Rule.ConditionCombiner.AND, Rule.Condition.of(FROM, NOT_CONTAINS, "AAA"));
+
+    private static Rule.Builder asRule(Action.Forward forward) {
+        return Rule.builder()
+            .id(Rule.Id.of("1"))
+            .name("rule 1")
+            .conditionGroup(CONDITION_GROUP)
+            .action(Action.builder().setForward(forward));
+    }
+
+    private TemporaryJamesServer jamesServer;
+    private FilteringManagementProbeImpl filteringManagementProbe;
+    private MailRepositoryProbeImpl mailRepositoryProbe;
+
+    @RegisterExtension
+    public SMTPMessageSender messageSender = new SMTPMessageSender(DEFAULT_DOMAIN);
+    private DataProbeImpl dataProbe;
+
+    @BeforeEach
+    void setup(@TempDir File temporaryFolder) throws Exception {
+        jamesServer = TemporaryJamesServer.builder()
+            .withOverrides(binder -> Multibinder.newSetBinder(binder, GuiceProbe.class).addBinding().to(FilteringManagementProbeImpl.class))
+            .withMailetContainer(MailetContainer.builder()
+                .putProcessor(ProcessorConfiguration.root()
+                    .addMailet(MailetConfiguration.builder()
+                        .matcher(All.class)
+                        .mailet(RecipientRewriteTable.class))
+                    .addMailet(MailetConfiguration.builder()
+                        .matcher(All.class)
+                        .mailet(JMAPFiltering.class))
+                    .addMailet(MailetConfiguration.builder()
+                        .matcher(All.class)
+                        .mailet(ToRepository.class)
+                        .addProperty("repositoryPath", CUSTOM_REPOSITORY.asString())))
+                .putProcessor(CommonProcessors.error())
+                .putProcessor(CommonProcessors.transport())
+                .putProcessor(CommonProcessors.bounces()))
+            .build(temporaryFolder);
+
+        jamesServer.start();
+
+        dataProbe = jamesServer.getProbe(DataProbeImpl.class);
+        dataProbe.addDomain(DEFAULT_DOMAIN);
+
+        dataProbe.addUser(SENDER.asString(), PASSWORD);
+        dataProbe.addUser(ALICE.asString(), PASSWORD);
+        dataProbe.addUser(BOB.asString(), PASSWORD);
+        dataProbe.addUser(CEDRIC.asString(), PASSWORD);
+
+        mailRepositoryProbe = jamesServer.getProbe(MailRepositoryProbeImpl.class);
+
+        filteringManagementProbe = jamesServer.getProbe(FilteringManagementProbeImpl.class);
+    }
+
+    @AfterEach
+    void tearDown() {
+        jamesServer.shutdown();
+    }
+
+    @Test
+    void filterForwardShouldNotCreateLoopError() throws Exception {
+        filteringManagementProbe.defineRulesForUser(ALICE, asRule(Forward.to(BOB.asMailAddress()).keepACopy()));
```

Comment:
I maintain a previous comment on this... Try honestly writing the same test  with alice doing a forward without a copy, and I would not be surprised that we just get no email being recorded in James in the end.

I think we agreed previously to store the mail at the end of the chain, and I don't see that here.

Reply:
fixed. In case A have to forward mail to B but B has already received the mail before, A will keep a local copy no matter what keepACopy is set